src/libkrun/src/lib.rs

```diff
@@ -1104,6 +1033,163 @@ fn create_virtio_net(ctx_cfg: &mut ContextConfig, backend: VirtioNetBackend) {
         .expect("Failed to create network interface");
 }
 
+#[cfg(all(not(feature = "efi"), not(feature = "tee")))]
+fn map_kernel(ctx_id: u32, kernel_path: &str) -> i32 {
+    let file = match File::options().read(true).write(false).open(kernel_path) {
+        Ok(file) => file,
+        Err(err) => {
+            error!("Error opening external kernel: {err}");
+            return -libc::EINVAL;
+        }
+    };
+
+    let kernel_size = file.metadata().unwrap().len();
+
+    let kernel_host_addr = unsafe {
+        libc::mmap(
+            std::ptr::null_mut(),
+            kernel_size as usize,
+            libc::PROT_READ,
+            libc::MAP_SHARED,
+            file.as_raw_fd(),
+            0_i64,
+        )
+    };
+    if kernel_host_addr == libc::MAP_FAILED {
+        error!("Can't load kernel into process map");
+        return -libc::EINVAL;
+    }
+
+    let kernel_bundle = KernelBundle {
+        host_addr: kernel_host_addr as u64,
+        guest_addr: 0x8000_0000,
+        entry_addr: 0x8000_0000,
+        size: kernel_size as usize,
+    };
+
+    match CTX_MAP.lock().unwrap().entry(ctx_id) {
+        Entry::Occupied(mut ctx_cfg) => ctx_cfg
+            .get_mut()
+            .vmr
+            .set_kernel_bundle(kernel_bundle)
+            .unwrap(),
+        Entry::Vacant(_) => return -libc::ENOENT,
+    }
+
+    KRUN_SUCCESS
+}
+
+#[cfg(any(feature = "tee", feature = "efi"))]
+#[allow(clippy::format_collect)]
+#[allow(clippy::missing_safety_doc)]
+#[no_mangle]
+pub unsafe extern "C" fn krun_set_kernel(_ctx_id: u32, _c_kernel_path: *const c_char) -> i32 {
+    -libc::EOPNOTSUPP
```

Comment:
What's holding us back from supporting `krun_set_kernel` with the `tee` feature?

Reply:
We still need custom kernel (at least for SEV/SNP, not so sure about TDX), a custom qboot (for the initial memory initialization), and a custom init (for bundling an agent and opening the LUKS volume).

We should get rid of those custom components. A possible plan for that could be:

- Adopt OVMF/EDK2 as FW. It would need to be a custom built, since we aren't 100% compatible with QEMU, but would have the facilities required to boot an unmodified kernel in a TEE.
- Find a way to generate an initramfs in a reproducible way, so it can be part of the attestation envelope.

Both tasks seem doable, but require a significant amount of effort. Once we get there, we'll be able to support external kernels with the TEE flavor.

Reply:
> * Find a way to generate an initramfs on the fly in a reproducible way, so it can be part of the attestation envelope.
> 

Can you expand on this a bit more?

Reply:
By the way, while it's most likely not the right tool for this, I tried to keep the initramfs bureaucracy as simple as possible in [mbuto](https://mbuto.sh/), so it might be helpful if you're trying to sketch stuff quickly.

Reply:
> By the way, while it's most likely not the right tool for this, I tried to keep the initramfs bureaucracy as simple as possible in [mbuto](https://mbuto.sh/), so it might be helpful if you're trying to sketch stuff quickly.

Hm... this may come handy for building initramfs to TEEs, thanks.